stdlib/SHA/src/sha2.jl

```diff
@@ -1,134 +1,162 @@
-            s0 = sigma0_256(s0)
-            s1 = unsafe_load(pbuf, mod1(j + 14, 16))
-            s1 = sigma1_256(s1)
-
-            # Apply the SHA-256 compression function to update a..h
-            v = unsafe_load(pbuf, mod1(j, 16)) + s1 + unsafe_load(pbuf, mod1(j + 9, 16)) + s0
-            unsafe_store!(pbuf, v, mod1(j, 16))
-            T1 = h + Sigma1_256(e) + Ch(e, f, g) + K256[j] + v
-            T2 = Sigma0_256(a) + Maj(a, b, c)
-            h = g
-            g = f
-            f = e
-            e = UInt32(d + T1)
-            d = c
-            c = b
-            b = a
-            a = UInt32(T1 + T2)
-        end
+    return esc(quote
+        # We byteswap every input byte
+        v = bswap(unsafe_load(pbuf, $j))
+        unsafe_store!(pbuf, v, $j)
+
+        # Apply the SHA-256 compression function to update a..h
+        T1 = $h + $Sigma1($e) + Ch($e, $f, $g) + $K[$j] + v
+        $h = $Sigma0($a) + Maj($a, $b, $c)
+        $d += T1
+        $h += T1
+    end)
+end
+
+macro R17_80(j, T)
+
+    ww = (:a, :b, :c, :d, :e, :f, :g, :h)
+
+    a = ww[((81 - j) % 8) + 1]
+    b = ww[((82 - j) % 8) + 1]
+    c = ww[((83 - j) % 8) + 1]
+    d = ww[((84 - j) % 8) + 1]
+    e = ww[((85 - j) % 8) + 1]
+    f = ww[((86 - j) % 8) + 1]
+    g = ww[((87 - j) % 8) + 1]
+    h = ww[((88 - j) % 8) + 1]
+
+    if T == 512
+        Sigma0 = :Sigma0_512
+        Sigma1 = :Sigma1_512
+        sigma0 = :sigma0_512
+        sigma1 = :sigma1_512
+        K = :K512
+    elseif T == 256
+        Sigma0 = :Sigma0_256
+        Sigma1 = :Sigma1_256
+        sigma0 = :sigma0_256
+        sigma1 = :sigma1_256
+        K = :K256
     end
 
-    # Compute the current intermediate hash value
-    context.state[1] += a
-    context.state[2] += b
-    context.state[3] += c
-    context.state[4] += d
-    context.state[5] += e
-    context.state[6] += f
-    context.state[7] += g
-    context.state[8] += h
+    return esc(quote
+        s0 = unsafe_load(pbuf, mod1($j + 1, 16))
+        s0 = $sigma0(s0)
+        s1 = unsafe_load(pbuf, mod1($j + 14, 16))
+        s1 = $sigma1(s1)
+
+        # Apply the SHA-256 compression function to update a..h
+        v = unsafe_load(pbuf, mod1($j, 16)) + s1 + unsafe_load(pbuf, mod1($j + 9, 16)) + s0
+        unsafe_store!(pbuf, v, mod1($j, 16))
+        T1 = $h + $Sigma1($e) + Ch($e, $f, $g) + $K[$j] + v
+        $h = $Sigma0($a) + Maj($a, $b, $c)
+        $d += T1
+        $h += T1
+    end)
 end
 
+function transform!(context::Union{SHA2_224_CTX,SHA2_256_CTX})
+    pbuf = buffer_pointer(context)
+    # Initialize registers with the previous intermediate values (our state)
+    a, b, c, d, e, f, g, h = context.state
+
+    # Initial Rounds
+    @R1_16( 1, 256); @R1_16( 2, 256); @R1_16( 3, 256); @R1_16( 4, 256);
+    @R1_16( 5, 256); @R1_16( 6, 256); @R1_16( 7, 256); @R1_16( 8, 256);
+    @R1_16( 9, 256); @R1_16(10, 256); @R1_16(11, 256); @R1_16(12, 256);
+    @R1_16(13, 256); @R1_16(14, 256); @R1_16(15, 256); @R1_16(16, 256);
```

Comment:
I'm always weary of lists of numbers like this. Maybe use another macro:
```
@static_unroll for i = 1:16
    @R1_16(i, 256)
end
```

Reply:
done, is there a `@static_unroll` macro?

Reply:
i did:
```julia
macro R_init(T)
    expr = :()
    for i in 1:16
        expr = :($expr; @R1_16($i, $T))
     end
    return esc(expr)
 end
```